webhooks/clusterclass_test.go

```diff
@@ -0,0 +1,155 @@
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+)
+
+var (
+	ctx        = ctrl.SetupSignalHandler()
+	fakeScheme = runtime.NewScheme()
+)
+
+func init() {
+	_ = clusterv1.AddToScheme(fakeScheme)
+}
+
+func TestClusterClassDefaultNamespaces(t *testing.T) {
+	// NOTE: ClusterTopology feature flag is disabled by default, thus preventing to create or update ClusterClasses.
+	// Enabling the feature flag temporarily for this test.
+	defer utilfeature.SetFeatureGateDuringTest(t, feature.Gates, feature.ClusterTopology, true)()
+
+	namespace := "default"
+	ref := &corev1.ObjectReference{
+		APIVersion: "foo",
+		Kind:       "barTemplate",
+		Name:       "baz",
+	}
+	in := &clusterv1.ClusterClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: namespace,
+		},
+		Spec: clusterv1.ClusterClassSpec{
+			Infrastructure: clusterv1.LocalObjectTemplate{Ref: ref},
+			ControlPlane: clusterv1.ControlPlaneClass{
+				LocalObjectTemplate:   clusterv1.LocalObjectTemplate{Ref: ref},
+				MachineInfrastructure: &clusterv1.LocalObjectTemplate{Ref: ref},
+			},
+			Workers: clusterv1.WorkersClass{
+				MachineDeployments: []clusterv1.MachineDeploymentClass{
+					{
+						Class: "aa",
+						Template: clusterv1.MachineDeploymentClassTemplate{
+							Bootstrap:      clusterv1.LocalObjectTemplate{Ref: ref},
+							Infrastructure: clusterv1.LocalObjectTemplate{Ref: ref},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	webhook := &ClusterClassWebhook{
+		Client: fake.NewClientBuilder().WithObjects().WithScheme(fakeScheme).Build(),
+	}
+
+	t.Run("for ClusterClass", utildefaulting.CustomDefaultValidateTest(ctx, in, webhook))
+
+	g := NewWithT(t)
+	g.Expect(webhook.Default(ctx, in)).To(Succeed())
+
+	// Namespace defaulted on references
+	g.Expect(in.Spec.Infrastructure.Ref.Namespace).To(Equal(namespace))
+	g.Expect(in.Spec.ControlPlane.Ref.Namespace).To(Equal(namespace))
+	g.Expect(in.Spec.ControlPlane.MachineInfrastructure.Ref.Namespace).To(Equal(namespace))
+	for i := range in.Spec.Workers.MachineDeployments {
+		g.Expect(in.Spec.Workers.MachineDeployments[i].Template.Bootstrap.Ref.Namespace).To(Equal(namespace))
+		g.Expect(in.Spec.Workers.MachineDeployments[i].Template.Infrastructure.Ref.Namespace).To(Equal(namespace))
+	}
+}
+
+func TestClusterClassValidateDelete(t *testing.T) {
+
+	clusterClass := &clusterv1.ClusterClass{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "my-clusterclass",
+			Namespace: "default",
+		},
+	}
+
+	tests := []struct {
+		name     string
+		clusters []client.Object
+		wantErr  bool
+	}{
+		{
+			name: "Fails if corresponding Cluster exists",
+			clusters: []client.Object{
+				builder.Cluster(metav1.NamespaceDefault, "my-cluster").
+					WithClusterClass(clusterClass).
+					Build(),
+			},
+			wantErr: true,
+		},
+		{
+			name:    "Succeeds if no corresponding Cluster exists",
+			wantErr: false,
+		},
```

Comment:
It's currently not possible to add a testcase with a Cluster using another ClusterClass as controller-runtime does not support field selector (https://github.com/kubernetes-sigs/controller-runtime/issues/1376).

Should I add a comment for that?